Cloud functions ingest: backfill CLI tool + minor improvements (#190)
GoogleCloudPlatform/bigquery-utils: tools/cloud_functions/gcs_event_based_ingest/README.md

Your regex can optionally include  for 
- `partition` must be BigQuery Partition decorator with leading `$`
- `yyyy`, `mm`, `dd`, `hr` partition year, month, day, and hour
(depending on your partition granularity)
- `batch` an optional batch id to indicate multiple uploads for this partition.

For example, if your datafiles were laid out like this:
```text
gs://${BUCKET}/${SOURCE_SYSTEM}/${DATASET}/${TABLE}/region=${LOCATION}/yyyy=${YEAR}/mm=${MONTH}/dd=${DAY}/hh=${HOUR}
```
i.e.
```text
gs://my-bucket/on-prem-edw/my_product/transactions/region=US/yyyy=2020/mm=01/dd=02/hh=03/_SUCCESS
```
Then you could use [this regex](https://regex101.com/r/OLpmg4/2):
```text
DESTINATION_REGEX='(?:[\w\-_0-9]+)/(?P<dataset>[\w\-_0-9]+)/(?P<table>[\w\-_0-9]+)/region=(?P<batch>[\w]+)/yyyy=(?P<yyyy>[0-9]{4})/mm=(?P<mm>[0-9]{2})/dd=(?P<dd>[0-9]{2})/hh=(?P<hh>[0-9]{2})/'
```
In this case we can take advantage of a more known rigid structure so our regex 
is simpler (no optional capturing groups, optional slashes).
Note, we can use the `region=` string (which may have been partitioned on
in an  upstream system such as Hive) as a batch ID because we might expect that

> an hourly partition might have mutliple directories that upload to it.

Reviewer: ```suggestion
an hourly partition might have multiple directories that upload to it.
```

(e.g. US, GB, etc). Because it is all named capturing groups we don't have any
strict ordering restrictions about batch id appearing before / after partition
information.

### Dealing with Different Naming Conventions in the Same Bucket

> In most cases, it would be recommeneded to have separate buckets / deployment 

Reviewer: ```suggestion
In most cases, it would be recommended to have separate buckets / deployment 
```

### Job Naming Convention

> All load or external query jobs will have a job id witha  prefix following this convention:

Reviewer: ```suggestion
All load or external query jobs will have a job id with a  prefix following this convention:
```

incur a small cost. These resources will be spun up based on your Google Cloud SDK
[Application Default Credentials](https://cloud.google.com/sdk/gcloud/reference/auth/application-default)

#### Pytest Fixtures
All Pytest fixtures are DRY-ed up into `tests/conftest.py`
This is mostly to share fixtures between the main integration test for the cloud function

> and the integrationt tests for the backfill CLI.

Reviewer: ```suggestion
and the integration tests for the backfill CLI.
```

utility to backfill success file notifications or run the cloud function locally in concurrent threads.

optional arguments:
  -h, --help            show this help message and exit
  --gcs-path GCS_PATH, -p GCS_PATH
                        GCS path (e.g. gs://bucket/prefix/to/search/)to search for existing _SUCCESS files
  --mode {LOCAL,NOTIFICATIONS}, -m {LOCAL,NOTIFICATIONS}
                        How to perform the backfill: LOCAL run cloud function main method locally (in concurrent threads) or NOTIFICATIONS just push notifications to Pub/Sub for a deployed
                        version of the cloud function to pick up. Default is NOTIFICATIONS.
  --pubsub-topic PUBSUB_TOPIC, --topic PUBSUB_TOPIC, -t PUBSUB_TOPIC
                        Pub/Sub notifications topic to post notifications for. i.e. projects/{PROJECT_ID}/topics/{TOPIC_ID} Required if using NOTIFICATIONS mode.
  --success-filename SUCCESS_FILENAME, -f SUCCESS_FILENAME
                        Override the default success filename '_SUCCESS'
  --destination-regex DESTINATION_REGEX, -r DESTINATION_REGEX

>                         Override the default destination regex for determining BigQuerydestination based on information encoded in the GCS path of thesuccess file

Reviewer: ```suggestion
                        Override the default destination regex for determining BigQuery destination based on information encoded in the GCS path of the success file
```
